Rework crosswalk

kirkkwang opened the pull request:
## Revert DC Crosswalk

f2a2f9be92ac5be225c1eaa46d441fa5994b9904

This commit will revert the DC Crosswalk changes made just after the
fork.  This wasn't the best implementation because not all the terms in
Hyku are Dublin ore.  Instead we will create a new crosswalk for Hyku.

## Introduce the HykuCrosswalk

a401c6b8e221c1f3657e775b2e3d5f5da7ef28b4

This commit will introduce the HykuCrosswalk class which will be
responsible for converting a work from Hyku to an Atom feed.  After
consulting with the client, this is now aligned with what they are
expecting the output to be.  We want to double the dc and dcterms
properties while falling back to dc even if Hyrax/Hyku uses the dcterm
predicate to better serve compatibility.  We've also chosen h4csys as
the namespace for system metadata that are largested automatically
generated.  We are using h4cmeta for any metadata that is settable by
the user.

Example of XML output:
```xml
<feed dc="http://purl.org/dc/elements/1.1/" dcterms="http://purl.org/dc/terms/" h4cmeta="https://hykucommons.org/schema/metadata" h4csys="https://hykucommons.org/schema/system">
  <title>A Test GenericWork, Testing for GenericWorkResource</title>
  <content rel="src" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b"/>
  <link rel="edit" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b/file_sets"/>
  <entry>
    <content rel="src" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b/file_sets/1998fc56-91cc-482e-ac5a-25e19a713da5"/>
    <link rel="edit" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b/file_sets/1998fc56-91cc-482e-ac5a-25e19a713da5"/>
  </entry>
  <h4csys:id>ff323cbf-b601-4236-8a7a-1b993e66d79b</h4csys:id>
  <h4csys:internal_resource>Image</h4csys:internal_resource>
  <h4csys:created_at>2024-07-10 23:04:45 UTC</h4csys:created_at>
  <h4csys:updated_at>2025-02-05 22:19:27 UTC</h4csys:updated_at>
  <h4csys:new_record>false</h4csys:new_record>
  <h4csys:date_modified>2025-02-03T23:30:24+00:00</h4csys:date_modified>
  <h4csys:date_uploaded>2025-02-03T23:30:24+00:00</h4csys:date_uploaded>
  <h4csys:depositor>admin@example.com</h4csys:depositor>
  <h4csys:state>http://fedora.info/definitions/1/0/access/ObjState#active</h4csys:state>
  <h4cmeta:embargo_id>ac7c756b-be04-4206-becd-fd8b4cee7636</h4cmeta:embargo_id>
  <h4cmeta:title>A Test GenericWork</h4cmeta:title>
  <h4cmeta:title>Testing for GenericWorkResource</h4cmeta:title>
  <h4cmeta:admin_set_id>0328531d-18d2-474c-a8ec-4cd6e052d519</h4cmeta:admin_set_id>
  <h4cmeta:member_ids>1998fc56-91cc-482e-ac5a-25e19a713da5</h4cmeta:member_ids>
  <h4cmeta:representative_id>1998fc56-91cc-482e-ac5a-25e19a713da5</h4cmeta:representative_id>
  <h4cmeta:thumbnail_id>1998fc56-91cc-482e-ac5a-25e19a713da5</h4cmeta:thumbnail_id>
  <h4cmeta:abstract>This is the abstract, it's so short!</h4cmeta:abstract>
  <h4cmeta:access_right>You can access it this.</h4cmeta:access_right>
  <h4cmeta:alternative_title>A testing of the highest calibre</h4cmeta:alternative_title>
  <h4cmeta:based_near>https://sws.geonames.org/5391811/</h4cmeta:based_near>
  <h4cmeta:bibliographic_citation>leave bibliographic citation here</h4cmeta:bibliographic_citation>
  <h4cmeta:contributor>Smith, Josh</h4cmeta:contributor>
  <h4cmeta:creator>Smith, John</h4cmeta:creator>
  <h4cmeta:creator>Smith, James</h4cmeta:creator>
  <h4cmeta:date_created>2020-07-07</h4cmeta:date_created>
  <h4cmeta:description>This is the description, much longer than the abstract, but not by much in this case.</h4cmeta:description>
  <h4cmeta:identifier>abc123</h4cmeta:identifier>
  <h4cmeta:keyword>test</h4cmeta:keyword>
  <h4cmeta:keyword>work</h4cmeta:keyword>
  <h4cmeta:publisher>Notch8</h4cmeta:publisher>
  <h4cmeta:label>test</h4cmeta:label>
  <h4cmeta:language>english</h4cmeta:language>
  <h4cmeta:license>https://creativecommons.org/licenses/by-nc/4.0/</h4cmeta:license>
  <h4cmeta:related_url>https://example.com/</h4cmeta:related_url>
  <h4cmeta:resource_type>Audio</h4cmeta:resource_type>
  <h4cmeta:resource_type>Capstone Project</h4cmeta:resource_type>
  <h4cmeta:rights_notes>These are the rights notes</h4cmeta:rights_notes>
  <h4cmeta:rights_statement>http://rightsstatements.org/vocab/NoC-CR/1.0/</h4cmeta:rights_statement>
  <h4cmeta:source>From the internet</h4cmeta:source>
  <h4cmeta:subject>WillowSword</h4cmeta:subject>
  <h4cmeta:subject>testing</h4cmeta:subject>
  <h4cmeta:extent>50 mins</h4cmeta:extent>
  <h4cmeta:video_embed>https://www.youtube.com/embed/Znf73dsFdC8</h4cmeta:video_embed>
  <h4cmeta:visibility_during_embargo>restricted</h4cmeta:visibility_during_embargo>
  <h4cmeta:visibility_after_embargo>open</h4cmeta:visibility_after_embargo>
  <h4cmeta:embargo_release_date>2027-10-04T00:00:00+00:00</h4cmeta:embargo_release_date>
  <dc:title>A Test GenericWork</dc:title>
  <dc:title>Testing for GenericWorkResource</dc:title>
  <dcterms:abstract>This is the abstract, it's so short!</dcterms:abstract>
  <dcterms:accessRights>You can access it this.</dcterms:accessRights>
  <dcterms:alternative>A testing of the highest calibre</dcterms:alternative>
  <dcterms:bibliographicCitation>leave bibliographic citation here</dcterms:bibliographicCitation>
  <dc:contributor>Smith, Josh</dc:contributor>
  <dc:creator>Smith, John</dc:creator>
  <dc:creator>Smith, James</dc:creator>
  <dc:date>2020-07-07</dc:date>
  <dc:description>This is the description, much longer than the abstract, but not by much in this case.</dc:description>
  <dc:identifier>abc123</dc:identifier>
  <dc:publisher>Notch8</dc:publisher>
  <dc:language>english</dc:language>
  <dcterms:license>https://creativecommons.org/licenses/by-nc/4.0/</dcterms:license>
  <dc:type>Audio</dc:type>
  <dc:type>Capstone Project</dc:type>
  <dc:rights>These are the rights notes</dc:rights>
  <dc:rights>http://rightsstatements.org/vocab/NoC-CR/1.0/</dc:rights>
  <dc:source>From the internet</dc:source>
  <dc:subject>WillowSword</dc:subject>
  <dc:subject>testing</dc:subject>
  <dcterms:extent>50 mins</dcterms:extent>
  <dcterms:modified>2025-02-03T23:30:24+00:00</dcterms:modified>
  <dcterms:dateSubmitted>2025-02-03T23:30:24+00:00</dcterms:dateSubmitted>
</feed>
```

## Add type to collection show

cab14d972d5ba53c685e4e2974193eb5a8112ec0

This commit will add the collection type (Collection or AdminSet) to the
feed.

## Add adjust and add specs

a3e096c6b8b13b4b8f52b013b68c1555d035a94d

This commit will adjust the dc_crosswalk_spec and also add specs for the
hyku_crosswalk as well as the show.hyku.xml.builder.  Currently, it was
only tested on Ruby 2.7.6.

Commit:
Revert DC Crosswalk

This commit will revert the DC Crosswalk changes made just after the
fork.  This wasn't the best implementation because not all the terms in
Hyku are Dublin ore.  Instead we will create a new crosswalk for Hyku.

Commit:
Introduce the HykuCrosswalk

This commit will introduce the HykuCrosswalk class which will be
responsible for converting a work from Hyku to an Atom feed.  After
consulting with the client, this is now aligned with what they are
expecting the output to be.  We want to double the dc and dcterms
properties while falling back to dc even if Hyrax/Hyku uses the dcterm
predicate to better serve compatibility.  We've also chosen h4csys as
the namespace for system metadata that are largested automatically
generated.  We are using h4cmeta for any metadata that is settable by
the user.

Example of XML output:
```xml
<feed dc="http://purl.org/dc/elements/1.1/" dcterms="http://purl.org/dc/terms/" h4cmeta="https://hykucommons.org/schema/metadata" h4csys="https://hykucommons.org/schema/system">
  <title>A Test GenericWork, Testing for GenericWorkResource</title>
  <content rel="src" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b"/>
  <link rel="edit" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b/file_sets"/>
  <entry>
    <content rel="src" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b/file_sets/1998fc56-91cc-482e-ac5a-25e19a713da5"/>
    <link rel="edit" href="http://dev.hyku.test/sword/collections/default/works/ff323cbf-b601-4236-8a7a-1b993e66d79b/file_sets/1998fc56-91cc-482e-ac5a-25e19a713da5"/>
  </entry>
  <h4csys:id>ff323cbf-b601-4236-8a7a-1b993e66d79b</h4csys:id>
  <h4csys:internal_resource>Image</h4csys:internal_resource>
  <h4csys:created_at>2024-07-10 23:04:45 UTC</h4csys:created_at>
  <h4csys:updated_at>2025-02-05 22:19:27 UTC</h4csys:updated_at>
  <h4csys:new_record>false</h4csys:new_record>
  <h4csys:date_modified>2025-02-03T23:30:24+00:00</h4csys:date_modified>
  <h4csys:date_uploaded>2025-02-03T23:30:24+00:00</h4csys:date_uploaded>
  <h4csys:depositor>admin@example.com</h4csys:depositor>
  <h4csys:state>http://fedora.info/definitions/1/0/access/ObjState#active</h4csys:state>
  <h4cmeta:embargo_id>ac7c756b-be04-4206-becd-fd8b4cee7636</h4cmeta:embargo_id>
  <h4cmeta:title>A Test GenericWork</h4cmeta:title>
  <h4cmeta:title>Testing for GenericWorkResource</h4cmeta:title>
  <h4cmeta:admin_set_id>0328531d-18d2-474c-a8ec-4cd6e052d519</h4cmeta:admin_set_id>
  <h4cmeta:member_ids>1998fc56-91cc-482e-ac5a-25e19a713da5</h4cmeta:member_ids>
  <h4cmeta:representative_id>1998fc56-91cc-482e-ac5a-25e19a713da5</h4cmeta:representative_id>
  <h4cmeta:thumbnail_id>1998fc56-91cc-482e-ac5a-25e19a713da5</h4cmeta:thumbnail_id>
  <h4cmeta:abstract>This is the abstract, it's so short!</h4cmeta:abstract>
  <h4cmeta:access_right>You can access it this.</h4cmeta:access_right>
  <h4cmeta:alternative_title>A testing of the highest calibre</h4cmeta:alternative_title>
  <h4cmeta:based_near>https://sws.geonames.org/5391811/</h4cmeta:based_near>
  <h4cmeta:bibliographic_citation>leave bibliographic citation here</h4cmeta:bibliographic_citation>
  <h4cmeta:contributor>Smith, Josh</h4cmeta:contributor>
  <h4cmeta:creator>Smith, John</h4cmeta:creator>
  <h4cmeta:creator>Smith, James</h4cmeta:creator>
  <h4cmeta:date_created>2020-07-07</h4cmeta:date_created>
  <h4cmeta:description>This is the description, much longer than the abstract, but not by much in this case.</h4cmeta:description>
  <h4cmeta:identifier>abc123</h4cmeta:identifier>
  <h4cmeta:keyword>test</h4cmeta:keyword>
  <h4cmeta:keyword>work</h4cmeta:keyword>
  <h4cmeta:publisher>Notch8</h4cmeta:publisher>
  <h4cmeta:label>test</h4cmeta:label>
  <h4cmeta:language>english</h4cmeta:language>
  <h4cmeta:license>https://creativecommons.org/licenses/by-nc/4.0/</h4cmeta:license>
  <h4cmeta:related_url>https://example.com/</h4cmeta:related_url>
  <h4cmeta:resource_type>Audio</h4cmeta:resource_type>
  <h4cmeta:resource_type>Capstone Project</h4cmeta:resource_type>
  <h4cmeta:rights_notes>These are the rights notes</h4cmeta:rights_notes>
  <h4cmeta:rights_statement>http://rightsstatements.org/vocab/NoC-CR/1.0/</h4cmeta:rights_statement>
  <h4cmeta:source>From the internet</h4cmeta:source>
  <h4cmeta:subject>WillowSword</h4cmeta:subject>
  <h4cmeta:subject>testing</h4cmeta:subject>
  <h4cmeta:extent>50 mins</h4cmeta:extent>
  <h4cmeta:video_embed>https://www.youtube.com/embed/Znf73dsFdC8</h4cmeta:video_embed>
  <h4cmeta:visibility_during_embargo>restricted</h4cmeta:visibility_during_embargo>
  <h4cmeta:visibility_after_embargo>open</h4cmeta:visibility_after_embargo>
  <h4cmeta:embargo_release_date>2027-10-04T00:00:00+00:00</h4cmeta:embargo_release_date>
  <dc:title>A Test GenericWork</dc:title>
  <dc:title>Testing for GenericWorkResource</dc:title>
  <dcterms:abstract>This is the abstract, it's so short!</dcterms:abstract>
  <dcterms:accessRights>You can access it this.</dcterms:accessRights>
  <dcterms:alternative>A testing of the highest calibre</dcterms:alternative>
  <dcterms:bibliographicCitation>leave bibliographic citation here</dcterms:bibliographicCitation>
  <dc:contributor>Smith, Josh</dc:contributor>
  <dc:creator>Smith, John</dc:creator>
  <dc:creator>Smith, James</dc:creator>
  <dc:date>2020-07-07</dc:date>
  <dc:description>This is the description, much longer than the abstract, but not by much in this case.</dc:description>
  <dc:identifier>abc123</dc:identifier>
  <dc:publisher>Notch8</dc:publisher>
  <dc:language>english</dc:language>
  <dcterms:license>https://creativecommons.org/licenses/by-nc/4.0/</dcterms:license>
  <dc:type>Audio</dc:type>
  <dc:type>Capstone Project</dc:type>
  <dc:rights>These are the rights notes</dc:rights>
  <dc:rights>http://rightsstatements.org/vocab/NoC-CR/1.0/</dc:rights>
  <dc:source>From the internet</dc:source>
  <dc:subject>WillowSword</dc:subject>
  <dc:subject>testing</dc:subject>
  <dcterms:extent>50 mins</dcterms:extent>
  <dcterms:modified>2025-02-03T23:30:24+00:00</dcterms:modified>
  <dcterms:dateSubmitted>2025-02-03T23:30:24+00:00</dcterms:dateSubmitted>
</feed>
```

Commit:
Add type to collection show

This commit will add the collection type (Collection or AdminSet) to the
feed.

ShanaLMoore reviewed (commented):
Great work sticking with this and refining the solution, @kirkkwang! Your persistence really shows.  

Before merging, could you add specs for the `HykuCrossWalk` class to ensure future stability? Should the `DcCrossWalk` specs also be updated? I’d expect changing the `initialize` method signature to cause failures, for example. Also, would it make sense to update the README to clarify expectations for the client (and future us)?  

Thanks again! 🚀

kirkkwang commented:
@ShanaLMoore Sounds good thanks, I'll work on specs and check the DcCrosswalk one.

Commit:
Add adjust and add specs

This commit will adjust the dc_crosswalk_spec and also add specs for the
hyku_crosswalk as well as the show.hyku.xml.builder.  Currently, it was
only tested on Ruby 2.7.6.

kirkkwang commented:
@ShanaLMoore ready for another look

ShanaLMoore reviewed (approved):
nice work! 🚀
